### ISSUE TITLE
Prepare the OpenSSL 1.1 support

### ISSUE THREAD

**Issue by dereckson** [Antti/rust-amqp#43]
OpenSSL 1.1 is coming on packages, for example it landed on Debian Sid next week.

That calls to bump the `openssl` crate dependency to 0.9.

As the `openssl` 0.7 create doesn't support OpensSL 1.1, as soon as an operating system or a Linux distribution bumps to OpenSSL 1.1, it becomes impossible to compile the `amqp` crate.

The upgrade isn't straightforward, as both the `openssl` crate and OpenSSL API changed, mostly:
* [Error handling has been modified](https://github.com/sfackler/rust-openssl/commit/fa622326490e1dd27df4d42b4097ca574deedb3)
* In OpenSSL 1.1, TLS protocol is now `TLS_method` instead of `SSLv23_method`, so SslMethod::Sslv23 dosn't exist anymore in `openssl` crate

**Comment by Antti** [comment]
At the moment it's not straight forward to do, because we clone connection for a reading loop and for channel writes. Since openssl v0.9, the `try_clone` method was removed from the TcpStream wrapper. So now it's impossible to do a concurrent writes/reads from different threads.

**Comment by Keruspe** [comment]
@Antti Is there any progress on this? Do you need help?

**Comment by Antti** [comment]
@Keruspe the biggest issue here is that new `SslStream` does not implement `clone` or `try_clone`. Thus making multithreaded reads and writes impossible.
I need to completely rewrite frames reading/dispatching code to make it work.

**Comment by kpcyrd** [comment]
Any news? I've used the amqp crate in one of my projects and I'm not able to compile it anymore because it's pulling in openssl 0.7

**Comment by steve-chavez** [comment]
Need as well, after an Archlinux upgrade I'm unable to compile the lib.

**Comment by grahamc** [comment]
@Antti I wonder if I could wrap the connection in a Arc<Mutex<...>> and share the writes with that?

**Comment by grahamc** [comment]
(without having to rebuild everything else ...) if you're available on IRC, ping me: gchristensen (freenode, mozilla.)

**Comment by grahamc** [comment]
I have an implementation using the openssl crate version 0.9.x here: https://github.com/Antti/rust-amqp/pull/59

**Comment by Antti** [comment]
@grahamc Mutex won't work, since it will lock socket for reading as well and will result in the deadlock.

**Comment by grahamc** [comment]
Yeah, I found that to be true quite quickly :( I used a different approach.

**Comment by whitfin** [comment]
Not to bug, but is there any movement on this that we can maybe assist on? Currently not possible to use this library alongside Hyper due to clashing `openssl` crates and neither is compatible with the other.
